### PR TITLE
Issue 18076 front end login

### PR DESCRIPTION
**Pull request by wezell**
This PR does a number of things:

It unifies the variables and tooling around front-end users and back-end users. dotCMS used to use 2 session variables to store users - USER and cms.user. Now, only USER is used. To get your front end user in velocity, you would do either $user or worse you could do $session.getAttribute("USER")
A user is a user. The only difference between a front end user and a back end user is if they have been set as "front end" or "back end" users in the admin.
Front end users can only use PageMode.LIVE. If a user is a front-end user, they are not allowed to specify or set a page mode that is not LIVE, meaning, front end users cannot preview unpublished content, pages etc
This also fixes an issue that prevented failed login attempts from being wiped out on a successful login - which would lead to long delays when logging in.

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
+package com.dotmarketing.util;
+
+import static org.junit.Assert.*;
+import javax.servlet.http.HttpServletRequest;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import com.dotcms.datagen.UserDataGen;
+import com.dotcms.mock.request.BaseRequest;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java]: Codacy Issue found: [Avoid unused imports such as 'com.dotcms.mock.request.BaseRequest'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+     * returns CMS_ANON if not set
+     * 
+     */
+    @Test
+    public void test_getUser() {
+
+        assert (userWebAPI.getUser(backEndRequest()).equals(backEndUser));
+        assert (userWebAPI.getUser(frontEndRequest()).equals(frontEndUser));
+
+        assert (userWebAPI.getUser(anonymousRequest()).equals(userWebAPI.getAnonymousUserNoThrow()));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import com.dotcms.util.IntegrationTestInitService;
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.UserAPI;
+import com.liferay.portal.model.User;
+import com.liferay.portal.util.WebKeys;
+
+public class PageModeTest {
+
+    private static User systemUser;
+    private static User testUser;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java]: Codacy Issue found: [Avoid unused private fields such as 'testUser'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+     * validates that isLoggedToBackend true if front end or back end OR returns CMS_ANON if not set
+     * 
+     */
+    @Test
+    public void test_isLoggedToBackend() {
+
+        assert (userWebAPI.isLoggedToBackend(backEndRequest()) == true);
+        assert (userWebAPI.isLoggedToBackend(frontEndRequest()) == false);
+
+        assert (userWebAPI.isLoggedToBackend(anonymousRequest()) == false);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Avoid unnecessary comparisons in boolean expressions](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+
+    /**
+     * Validates that only front end users can be returned from the call to
+     * userWebAPI.getLoggedInFrontendUser.
+     */
+    @Test
+    public void test_getLoggedInFrontendUser() {
+
+        assert (userWebAPI.getLoggedInFrontendUser(frontEndRequest()).equals(frontEndUser));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/main/java/com/dotmarketing/business/web/UserWebAPIImpl.java [hunk 6]
```diff
@@ -16,55 +13,65 @@
...
+    public User getLoggedInFrontendUser(HttpServletRequest request) {
+        User user = getLoggedInUser(request);
+
+        return user != null && PortalUtil.getUser(request).isFrontendUser() ? user : null;
+    }
+    
+    @CloseDBIfOpened
+    @Override
+    public boolean isLoggedToFrontend(HttpServletRequest request) {
+        User user = getLoggedInUser(request);
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/web/UserWebAPIImpl.java]: Codacy Issue found: [Local variable 'user' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+
+    /**
+     * Front end users can never have any page mode other than LIVE, even when you try to set it
+     * explictily
+     */
+
+    @Test
+    public void test_front_end_user_is_only_LIVE() {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * validates that getUser returns any user that is logged into the dotCMS front end or back end OR
+     * returns CMS_ANON if not set
+     * 
+     */
+    @Test
+    public void test_isLoggedToFrontend() {
+
+        assert (userWebAPI.isLoggedToFrontend(backEndRequest()) == false);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    private HttpServletRequest backEndRequest() {
+        final HttpServletRequest request = anonymousRequest();
+        assertTrue("backEndUser has backend role", backEndUser.isBackendUser());
+        request.setAttribute(WebKeys.USER, backEndUser);
+        return request;
+    }
+
+
+    @BeforeClass
+    public static void prepare() throws Exception {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [A method/constructor should not explicitly throw java.lang.Exception](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+import com.dotcms.mock.request.MockHttpRequest;
+import com.dotcms.util.IntegrationTestInitService;
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.UserAPI;
+import com.liferay.portal.model.User;
+import com.liferay.portal.util.WebKeys;
+
+public class PageModeTest {
+
+    private static User systemUser;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java]: Codacy Issue found: [Avoid unused private fields such as 'systemUser'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+     * 
+     */
+    @Test
+    public void test_getLoggedInUser() {
+
+        assert (userWebAPI.getLoggedInUser(frontEndRequest()).equals(frontEndUser));
+        assert (userWebAPI.getLoggedInUser(backEndRequest()).equals(backEndUser));
+
+
+        assert (userWebAPI.getLoggedInUser(anonymousRequest()) == null);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+     * returns CMS_ANON if not set
+     * 
+     */
+    @Test
+    public void test_isLoggedToFrontend() {
+
+        assert (userWebAPI.isLoggedToFrontend(backEndRequest()) == false);
+        assert (userWebAPI.isLoggedToFrontend(frontEndRequest()) == true);
+
+        assert (userWebAPI.isLoggedToFrontend(anonymousRequest()) == false);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+        assert (PageMode.get(request) == PageMode.EDIT_MODE);
+
+    }
+
+    /**
+     * getting and setting a PAGE MODE does not create a session
+     */
+    @Test
+    public void test_page_mode_does_not_create_session() {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/main/java/com/dotmarketing/business/web/UserWebAPIImpl.java [hunk 14]
```diff
@@ -16,55 +13,65 @@
...
+
+        return (user != null && !user.isAnonymousUser() && user.isActive() && (user.isBackendUser() || user.isFrontendUser()))
+                        ? user
+                        : null;
+    }
+
+    @Deprecated
+    @Override
+    public User getLoggedInUser(final HttpSession session) {
+        User user = PortalUtil.getUser(session);
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/web/UserWebAPIImpl.java]: Codacy Issue found: [Local variable 'user' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+     * Validates that only front end users can be returned from the call to
+     * userWebAPI.getLoggedInFrontendUser.
+     */
+    @Test
+    public void test_getLoggedInFrontendUser() {
+
+        assert (userWebAPI.getLoggedInFrontendUser(frontEndRequest()).equals(frontEndUser));
+        assert (userWebAPI.getLoggedInFrontendUser(backEndRequest()) == null);
+
+        assert (userWebAPI.getLoggedInFrontendUser(anonymousRequest()) == null);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+import com.dotmarketing.business.APILocator;
+import com.dotmarketing.business.UserAPI;
+import com.liferay.portal.model.User;
+import com.liferay.portal.util.WebKeys;
+
+public class PageModeTest {
+
+    private static User systemUser;
+    private static User testUser;
+    private static UserAPI userAPI;
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java]: Codacy Issue found: [Avoid unused private fields such as 'userAPI'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+        assert (userWebAPI.getUser(anonymousRequest()).equals(userWebAPI.getAnonymousUserNoThrow()));
+
+    }
+
+    /**
+     * validates that isLoggedToBackend true if front end or back end OR returns CMS_ANON if not set
+     * 
+     */
+    @Test
+    public void test_isLoggedToBackend() {
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [JUnit tests should include assert() or fail()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+    }
+
+    /**
+     * getting and setting a PAGE MODE does not create a session
+     */
+    @Test
+    public void test_page_mode_does_not_create_session() {
+
+        HttpServletRequest request = backEndRequest();
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/util/PageModeTest.java]: Codacy Issue found: [Local variable 'request' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * validates that getLoggedInUser returns any user that is logged into the dotCMS front end or back
+     * end
+     * 
+     */
+    @Test
+    public void test_getLoggedInUser() {
+
+        assert (userWebAPI.getLoggedInUser(frontEndRequest()).equals(frontEndUser));
+        assert (userWebAPI.getLoggedInUser(backEndRequest()).equals(backEndUser));
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java (hunk 8 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Avoid unnecessary comparisons in boolean expressions](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * validates that isLoggedToBackend true if front end or back end OR returns CMS_ANON if not set
+     * 
+     */
+    @Test
+    public void test_isLoggedToBackend() {
+
+        assert (userWebAPI.isLoggedToBackend(backEndRequest()) == true);
+        assert (userWebAPI.isLoggedToBackend(frontEndRequest()) == false);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Avoid unnecessary comparisons in boolean expressions](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * validates that getUser returns any user that is logged into the dotCMS front end or back end OR
+     * returns CMS_ANON if not set
+     * 
+     */
+    @Test
+    public void test_isLoggedToFrontend() {
+
+        assert (userWebAPI.isLoggedToFrontend(backEndRequest()) == false);
+        assert (userWebAPI.isLoggedToFrontend(frontEndRequest()) == true);
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Codacy Issue found: [Avoid unnecessary comparisons in boolean expressions](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/main/java/com/dotmarketing/util/PageMode.java [hunk 22]
```diff
@@ -166,9 +177,15 @@ private static PageMode getCurrentPageMode(final HttpSession ses) {
     }
 
     private static boolean isNavigateEditMode(final HttpSession ses) {
-        PageMode sessionPageMode = (PageMode) ses.getAttribute(WebKeys.PAGE_MODE_SESSION);
+        PageMode sessionPageMode = ses==null ? DEFAULT_PAGE_MODE : (PageMode) ses.getAttribute(WebKeys.PAGE_MODE_SESSION);
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/util/PageMode.java]: Codacy Issue found: [Local variable 'sessionPageMode' could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

File: dotCMS/src/main/java/com/dotmarketing/business/web/UserWebAPIImpl.java [hunk 23]
```diff
@@ -16,55 +13,65 @@
...
+        return (user != null && !user.isAnonymousUser() && user.isActive() && (user.isBackendUser() || user.isFrontendUser()))
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/web/UserWebAPIImpl.java]: Codacy Issue found: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5106108)

**Review by jgambarios — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+
+        assert (userWebAPI.isLoggedToBackend(anonymousRequest()) == false);
+
+    }
+
+
+    /**
+     * validates that getUser returns any user that is logged into the dotCMS front end or back end OR
+     * returns CMS_ANON if not set
+     *
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotmarketing/business/web/UserWebAPIImplTest.java]: Looks like a wrong javadoc here

**Review by dsilvam — APPROVED** [review]

**Pull request merged**